rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -50,115 +53,83 @@ public ObjectPoolPartition(ObjectPool<T> pool, PoolConfig config,
+          objectQueue.add(new Poolable<>(object, pool, host));
+          totalCount.incrementAndGet();
       }
     }
+    catch (Exception e) {
+      // skipping logging the exception as factories are already logging.
+    }
   }
 
   public void returnObject(Poolable<T> object)
   {
-    if (!objectFactory.validate(object.getObject())) {
-      log.debug(String.format("Invalid object...removing: %s ", object));
-      decreaseObject(object);
-      // Compensate for the removed object. Needed to prevent endless wait when in parallel a borrowObject is called
-      increaseObjects(1, false);
-      return;
-    }
+    try {
+      if (!objectFactory.validate(object.getObject())) {
+        log.debug(String.format("Invalid object...removing: %s ", object));
+        decreaseObject(object);
+        return;
+      }
 
-    log.debug(String.format("Returning object: %s to queue. Queue size: %d", object, objectQueue.size()));
-    if (!objectQueue.offer(object)) {
-      log.warn("Created more objects than configured. Created=" + totalCount + " QueueSize=" + objectQueue.size());
-      decreaseObject(object);
+      log.debug(String.format("Returning object: %s to queue. Queue size: %d", object, objectQueue.size()));
+      if (!objectQueue.offer(object)) {
+        log.warn("Created more objects than configured. Created=" + totalCount + " QueueSize=" + objectQueue.size());
+        decreaseObject(object);
+      }
+    }
+    finally {
+      takeSemaphore.release();
     }
   }
 
   public Poolable<T> getObject(boolean blocking)
   {
-    if (objectQueue.size() == 0) {
-      // increase objects and return one, it will return null if pool reaches max size or if object creation fails
-      Poolable<T> object = increaseObjects(this.config.getDelta(), true);
-
-      if (object != null) {
-        return object;
-      }
-
-      if (totalCount == 0) {
-        // Could not create objects, this is mostly due to connection timeouts hence no point blocking as there is not other producer of sockets
-        throw new RuntimeException("Could not add connections to pool");
-      }
-      // else wait for a connection to get free
-    }
-
-    Poolable<T> freeObject;
+    Poolable<T> object;
     try {
       if (blocking) {
-        freeObject = objectQueue.take();
+        takeSemaphore.acquire();
+        object = objectQueue.take();
       }
       else {
-        freeObject = objectQueue.poll(config.getMaxWaitMilliseconds(), TimeUnit.MILLISECONDS);
-        if (freeObject == null) {
-          throw new RuntimeException("Cannot get a free object from the pool");
-        }
+        takeSemaphore.tryAcquire(config.getMaxWaitMilliseconds(), TimeUnit.MILLISECONDS);
+        object = tryGetObject();
       }
+      object.setLastAccessTs(System.currentTimeMillis());
     }
-    catch (InterruptedException e) {
-      throw new RuntimeException(e); // will never happen
+    catch (Exception e) {
+      takeSemaphore.release();
```

Comment:
is it ok to release the semaphore even if the thread was interrupted while trying to acquire the lock from semaphore?

Reply:
It is not. We should call release only when we acquire a permit

Reply:
yeah, this needs to be fixed. will fix it in the next iteration.

Reply:
fixed.